File: gui/preview_overlay.py
from PyQt5.QtWidgets import QWidget, QApplication, QDesktopWidget
from PyQt5.QtCore import Qt, QTimer, QRect, QPoint
from PyQt5.QtGui import QPainter, QColor, QPen, QBrush, QFont, QGuiApplication, QScreen
from typing import Optional, Tuple, List
import os


class PreviewOverlay(QWidget):
    def __init__(self, duration: int = 2000):
        super().__init__()
        self._duration = duration
        self._preview_type = None
        self._preview_data = {}
        
        self.setWindowFlags(
            Qt.FramelessWindowHint |
            Qt.WindowStaysOnTopHint |
            Qt.Tool |
            Qt.WindowTransparentForInput
        )
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.setAttribute(Qt.WA_ShowWithoutActivating)
        
        self._timer = QTimer(self)
        self._timer.timeout.connect(self._on_timer)
        self._blink_state = True
        self._blink_count = 0
        
        self._setup_geometry()
    
    def _setup_geometry(self):
        screens = QGuiApplication.screens()
        if screens:
            virtual_geometry = screens[0].geometry()
            for screen in screens[1:]:
                virtual_geometry = virtual_geometry.united(screen.geometry())
            self.setGeometry(virtual_geometry)
        else:
            self.showFullScreen()
    
    def show_click_position(self, x: int, y: int, label: str = ""):
        self._preview_type = 'click'
        self._preview_data = {'x': x, 'y': y, 'label': label}
        self._start_preview()
    
    def show_drag_line(self, start_x: int, start_y: int, end_x: int, end_y: int):
        self._preview_type = 'drag'
        self._preview_data = {
            'start_x': start_x, 'start_y': start_y,
            'end_x': end_x, 'end_y': end_y
        }
        self._start_preview()
    
    def show_region(self, x: int, y: int, width: int, height: int, label: str = ""):
        self._preview_type = 'region'
        self._preview_data = {'x': x, 'y': y, 'width': width, 'height': height, 'label': label}
        self._start_preview()
    
    def show_scroll_position(self, x: int, y: int, clicks: int):
        self._preview_type = 'scroll'
        self._preview_data = {'x': x, 'y': y, 'clicks': clicks}
        self._start_preview()
    
    def show_image_match(self, image_path: str, confidence: float = 0.9):
        self._preview_type = 'image'
        self._preview_data = {'image_path': image_path, 'confidence': confidence}
        self._start_preview()
    
    def show_text_preview(self, text: str, title: str = "文本预览"):
        self._preview_type = 'text'
        self._preview_data = {'text': text, 'title': title}
        self._start_preview()
    
    def show_hotkey_preview(self, keys: List[str]):
        self._preview_type = 'hotkey'
        self._preview_data = {'keys': keys}
        self._start_preview()
    
    def show_action_group_preview(self, group_name: str, action_count: int, description: str = ""):
        self._preview_type = 'action_group'
        self._preview_data = {
            'group_name': group_name,
            'action_count': action_count,
            'description': description
        }
        self._start_preview()
    
    def _start_preview(self):
        self._blink_state = True
        self._blink_count = 0
        self.show()
        self.raise_()
        self._timer.start(200)
    
    def _on_timer(self):
        self._blink_state = not self._blink_state
        self._blink_count += 1
        self.update()
        
        if self._blink_count >= self._duration // 200:
            self._timer.stop()
            self.hide()
    
    def stop_preview(self):
        self._timer.stop()
        self.hide()
    
    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.Antialiasing)
        
        if self._preview_type == 'click':
            self._draw_click_marker(painter)
        elif self._preview_type == 'drag':
            self._draw_drag_line(painter)
        elif self._preview_type == 'region':
            self._draw_region(painter)
        elif self._preview_type == 'scroll':
            self._draw_scroll_marker(painter)
        elif self._preview_type == 'image':
            self._draw_image_match(painter)
        elif self._preview_type == 'text':
            self._draw_text_preview(painter)
        elif self._preview_type == 'hotkey':
            self._draw_hotkey_preview(painter)
        elif self._preview_type == 'action_group':
            self._draw_action_group_preview(painter)
    
    def _draw_click_marker(self, painter: QPainter):
        x = self._preview_data.get('x', 0)
        y = self._preview_data.get('y', 0)
        label = self._preview_data.get('label', '')
        
        radius = 20
        color = QColor(255, 50, 50, 200) if self._blink_state else QColor(255, 100, 100, 150)
        
        painter.setPen(QPen(color, 3))
        painter.setBrush(QBrush(QColor(255, 50, 50, 80)))
        painter.drawEllipse(QPoint(x, y), radius, radius)
        
        painter.setPen(QPen(QColor(255, 50, 50), 2))
        painter.drawLine(x - 10, y, x + 10, y)
        painter.drawLine(x, y - 10, x, y + 10)
        
        if label:
            font = QFont()
            font.setPointSize(12)
            font.setBold(True)
            painter.setFont(font)
            painter.setPen(QPen(QColor(255, 255, 255)))
            
            text_rect = QRect(x + radius + 5, y - 15, 200, 30)
            painter.fillRect(text_rect, QColor(0, 0, 0, 180))
            painter.drawText(text_rect, Qt.AlignLeft | Qt.AlignVCenter, label)
    
    def _draw_drag_line(self, painter: QPainter):
        start_x = self._preview_data.get('start_x', 0)
        start_y = self._preview_data.get('start_y', 0)
        end_x = self._preview_data.get('end_x', 0)
        end_y = self._preview_data.get('end_y', 0)
        
        color = QColor(50, 150, 255, 200) if self._blink_state else QColor(100, 180, 255, 150)
        
        painter.setPen(QPen(color, 3))
        painter.drawLine(start_x, start_y, end_x, end_y)
        
        painter.setBrush(QBrush(color))
        
        dx = end_x - start_x
        dy = end_y - start_y
        length = (dx * dx + dy * dy) ** 0.5
        if length > 0:
            dx /= length
            dy /= length
            
            arrow_size = 15
            p1 = QPoint(end_x, end_y)
            p2 = QPoint(int(end_x - arrow_size * (dx + dy * 0.5)),
                       int(end_y - arrow_size * (dy - dx * 0.5)))
            p3 = QPoint(int(end_x - arrow_size * (dx - dy * 0.5)),
                       int(end_y - arrow_size * (dy + dx * 0.5)))
            
            painter.drawPolygon(p1, p2, p3)
        
        painter.setPen(QPen(QColor(50, 200, 50), 3))
        painter.setBrush(QBrush(QColor(50, 200, 50, 100)))
        painter.drawEllipse(QPoint(start_x, start_y), 10, 10)
        
        painter.setPen(QPen(QColor(255, 50, 50), 3))
        painter.setBrush(QBrush(QColor(255, 50, 50, 100)))
        painter.drawEllipse(QPoint(end_x, end_y), 10, 10)
        
        font = QFont()
        font.setPointSize(10)
        painter.setFont(font)
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        start_text = f"({start_x}, {start_y})"
        end_text = f"({end_x}, {end_y})"
        
        painter.fillRect(start_x - 40, start_y - 30, 80, 20, QColor(0, 0, 0, 180))
        painter.drawText(start_x - 40, start_y - 15, start_text)
        
        painter.fillRect(end_x - 40, end_y + 10, 80, 20, QColor(0, 0, 0, 180))
        painter.drawText(end_x - 40, end_y + 25, end_text)
    
    def _draw_region(self, painter: QPainter):
        x = self._preview_data.get('x', 0)
        y = self._preview_data.get('y', 0)
        width = self._preview_data.get('width', 100)
        height = self._preview_data.get('height', 100)
        label = self._preview_data.get('label', '')
        
        color = QColor(50, 200, 50, 200) if self._blink_state else QColor(100, 230, 100, 150)
        
        pen = QPen(color, 2, Qt.DashLine)
        painter.setPen(pen)
        painter.setBrush(QBrush(QColor(50, 200, 50, 30)))
        painter.drawRect(x, y, width, height)
        
        font = QFont()
        font.setPointSize(11)
        painter.setFont(font)
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        text = label if label else f"区域: {width}x{height}"
        text_rect = QRect(x, y - 25, len(text) * 10 + 20, 25)
        painter.fillRect(text_rect, QColor(50, 200, 50, 200))
        painter.drawText(text_rect, Qt.AlignCenter, text)
    
    def _draw_scroll_marker(self, painter: QPainter):
        x = self._preview_data.get('x', 0)
        y = self._preview_data.get('y', 0)
        clicks = self._preview_data.get('clicks', 0)
        
        color = QColor(255, 150, 50, 200) if self._blink_state else QColor(255, 180, 100, 150)
        
        painter.setPen(QPen(color, 3))
        painter.setBrush(QBrush(QColor(255, 150, 50, 80)))
        painter.drawEllipse(QPoint(x, y), 15, 15)
        
        direction = "↑" if clicks > 0 else "↓"
        abs_clicks = abs(clicks)
        
        font = QFont()
        font.setPointSize(14)
        font.setBold(True)
        painter.setFont(font)
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        painter.drawText(x - 7, y + 5, direction)
        
        text = f"滚动: {abs_clicks} 格 {'向上' if clicks > 0 else '向下'}"
        text_rect = QRect(x + 20, y - 15, 150, 30)
        painter.fillRect(text_rect, QColor(0, 0, 0, 180))
        painter.drawText(text_rect, Qt.AlignLeft | Qt.AlignVCenter, text)
    
    def _draw_image_match(self, painter: QPainter):
        image_path = self._preview_data.get('image_path', '')
        confidence = self._preview_data.get('confidence', 0.9)
        
        try:
            import pyautogui
            try:
                location = pyautogui.locateOnScreen(image_path, confidence=confidence)
            except pyautogui.ImageNotFoundException:
                location = None
            
            if location:
                color = QColor(50, 200, 50, 200) if self._blink_state else QColor(100, 230, 100, 150)
                
                pen = QPen(color, 3)
                painter.setPen(pen)
                painter.setBrush(QBrush(QColor(50, 200, 50, 50)))
                painter.drawRect(location.left, location.top, location.width, location.height)
                
                painter.setPen(QPen(color, 2))
                painter.drawLine(location.left, location.top, location.left + 15, location.top)
                painter.drawLine(location.left, location.top, location.left, location.top + 15)
                painter.drawLine(location.left + location.width, location.top, location.left + location.width - 15, location.top)
                painter.drawLine(location.left + location.width, location.top, location.left + location.width, location.top + 15)
                painter.drawLine(location.left, location.top + location.height, location.left + 15, location.top + location.height)
                painter.drawLine(location.left, location.top + location.height, location.left, location.top + location.height - 15)
                painter.drawLine(location.left + location.width, location.top + location.height, location.left + location.width - 15, location.top + location.height)
                painter.drawLine(location.left + location.width, location.top + location.height, location.left + location.width, location.top + location.height - 15)
                
                font = QFont()
                font.setPointSize(10)
                font.setBold(True)
                painter.setFont(font)
                painter.setPen(QPen(QColor(255, 255, 255)))
                
                line1 = f"✓ 匹配成功"
                line2 = f"位置: ({location.left}, {location.top})"
                line3 = f"尺寸: {location.width}x{location.height}"
                line4 = f"置信度: {confidence}"
                
                info_x = location.left
                info_y = location.top - 80
                if info_y < 10:
                    info_y = location.top + location.height + 10
                
                info_width = 160
                info_height = 75
                info_rect = QRect(info_x, info_y, info_width, info_height)
                painter.fillRect(info_rect, QColor(50, 200, 50, 230))
                
                painter.drawText(info_x + 10, info_y + 18, line1)
                painter.drawText(info_x + 10, info_y + 35, line2)
                painter.drawText(info_x + 10, info_y + 52, line3)
                painter.drawText(info_x + 10, info_y + 69, line4)
                
                center_x = location.left + location.width // 2
                center_y = location.top + location.height // 2
                painter.setPen(QPen(QColor(255, 255, 255), 2))
                painter.drawLine(center_x - 10, center_y, center_x + 10, center_y)
                painter.drawLine(center_x, center_y - 10, center_x, center_y + 10)
            else:
                font = QFont()
                font.setPointSize(12)
                font.setBold(True)
                painter.setFont(font)
                painter.setPen(QPen(QColor(255, 100, 100)))
                
                screen_center = self.rect().center()
                
                info_width = 200
                info_height = 60
                info_rect = QRect(
                    screen_center.x() - info_width // 2,
                    screen_center.y() - info_height // 2,
                    info_width,
                    info_height
                )
                painter.fillRect(info_rect, QColor(0, 0, 0, 220))
                
                painter.drawText(info_rect.adjusted(10, 10, -10, -30), Qt.AlignCenter, "✗ 未找到匹配图片")
                painter.setFont(QFont("", 10))
                painter.drawText(info_rect.adjusted(10, 35, -10, -10), Qt.AlignCenter, f"置信度: {confidence}")
        except Exception as e:
            font = QFont()
            font.setPointSize(12)
            painter.setFont(font)
            painter.setPen(QPen(QColor(255, 100, 100)))
            
            screen_center = self.rect().center()
            error_msg = str(e)[:50] if str(e) else "未知错误"
            text = f"预览错误: {error_msg}"
            text_rect = QRect(screen_center.x() - 200, screen_center.y() - 20, 400, 40)
            painter.fillRect(text_rect, QColor(0, 0, 0, 200))
            painter.drawText(text_rect, Qt.AlignCenter, text)
    
    def _draw_text_preview(self, painter: QPainter):
        text = self._preview_data.get('text', '')
        title = self._preview_data.get('title', '文本预览')
        
        screen_center = self.rect().center()
        
        font = QFont()
        font.setPointSize(12)
        painter.setFont(font)
        
        display_text = text[:100] + "..." if len(text) > 100 else text
        text_width = len(display_text) * 12 + 40
        text_height = 80
        
        text_rect = QRect(
            screen_center.x() - text_width // 2,
            screen_center.y() - text_height // 2,
            text_width,
            text_height
        )
        
        color = QColor(50, 100, 200, 230) if self._blink_state else QColor(70, 120, 220, 200)
        painter.fillRect(text_rect, color)
        
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        title_font = QFont()
        title_font.setPointSize(11)
        title_font.setBold(True)
        painter.setFont(title_font)
        painter.drawText(text_rect.adjusted(10, 10, -10, -text_height + 30), Qt.AlignLeft, title)
        
        content_font = QFont()
        content_font.setPointSize(10)
        painter.setFont(content_font)
        painter.drawText(text_rect.adjusted(10, 35, -10, -10), Qt.AlignLeft | Qt.TextWordWrap, display_text)
    
    def _draw_hotkey_preview(self, painter: QPainter):
        keys = self._preview_data.get('keys', [])
        
        screen_center = self.rect().center()
        
        font = QFont()
        font.setPointSize(14)
        font.setBold(True)
        painter.setFont(font)
        
        key_texts = []
        key_width = 0
        for key in keys:
            key_display = key.upper()
            if len(key_display) == 1:
                key_display = key_display
            key_texts.append(key_display)
            key_width += len(key_display) * 15 + 30
        
        total_width = key_width + 40
        total_height = 60
        
        bg_rect = QRect(
            screen_center.x() - total_width // 2,
            screen_center.y() - total_height // 2,
            total_width,
            total_height
        )
        
        color = QColor(100, 50, 200, 230) if self._blink_state else QColor(120, 70, 220, 200)
        painter.fillRect(bg_rect, color)
        
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        x_offset = bg_rect.left() + 20
        for i, key_text in enumerate(key_texts):
            key_rect = QRect(x_offset, bg_rect.top() + 15, len(key_text) * 15 + 20, 30)
            
            painter.fillRect(key_rect, QColor(255, 255, 255, 50))
            painter.drawRect(key_rect)
            painter.drawText(key_rect, Qt.AlignCenter, key_text)
            
            x_offset += key_rect.width() + 5
            
            if i < len(key_texts) - 1:
                plus_rect = QRect(x_offset, bg_rect.top() + 20, 15, 20)
                painter.drawText(plus_rect, Qt.AlignCenter, "+")
                x_offset += 20
    
    def _draw_action_group_preview(self, painter: QPainter):
        group_name = self._preview_data.get('group_name', '未知')
        action_count = self._preview_data.get('action_count', 0)
        description = self._preview_data.get('description', '')
        
        screen_center = self.rect().center()
        
        font = QFont()
        font.setPointSize(12)
        painter.setFont(font)
        
        title_text = f"📁 动作组: {group_name}"
        count_text = f"包含 {action_count} 个动作"
        desc_text = description[:50] + "..." if len(description) > 50 else description
        
        max_width = max(len(title_text) * 12, len(count_text) * 10, len(desc_text) * 10) + 60
        total_width = max(max_width, 300)
        total_height = 120 if description else 90
        
        bg_rect = QRect(
            screen_center.x() - total_width // 2,
            screen_center.y() - total_height // 2,
            total_width,
            total_height
        )
        
        color = QColor(76, 175, 80, 230) if self._blink_state else QColor(100, 200, 100, 200)
        painter.fillRect(bg_rect, color)
        
        painter.setPen(QPen(QColor(255, 255, 255)))
        
        title_font = QFont()
        title_font.setPointSize(14)
        title_font.setBold(True)
        painter.setFont(title_font)
        painter.drawText(bg_rect.adjusted(15, 15, -15, -total_height + 40), Qt.AlignLeft, title_text)
        
        content_font = QFont()
        content_font.setPointSize(11)
        painter.setFont(content_font)
        painter.drawText(bg_rect.adjusted(15, 45, -15, -total_height + 70), Qt.AlignLeft, count_text)
        
        if description:
            desc_font = QFont()
            desc_font.setPointSize(10)
            painter.setFont(desc_font)
            painter.setPen(QPen(QColor(220, 220, 220)))
            painter.drawText(bg_rect.adjusted(15, 75, -15, -10), Qt.AlignLeft | Qt.TextWordWrap, desc_text)

File: rust-workspace/gui/src/preview_overlay.cpp
#include "preview_overlay.h"

#include <QPainter>
#include <QPainterPath>
#include <QScreen>
#include <QGuiApplication>
#include <QApplication>
#include <QFontMetrics>
#include <QFileInfo>
#include <QPixmap>
#include <algorithm>
#include <cmath>

PreviewOverlay* PreviewOverlay::get_instance(QWidget* parent)
{
    static PreviewOverlay* instance = nullptr;
    if (!instance) {
        instance = new PreviewOverlay(parent);
    }
    return instance;
}

PreviewOverlay::PreviewOverlay(QWidget* parent)
    : QWidget(parent)
    , m_type(None)
    , m_scrollClicks(0)
    , m_confidence(0.0)
    , m_autoHideDuration(2000)
{
    setWindowFlags(Qt::FramelessWindowHint
                   | Qt::WindowStaysOnTopHint
                   | Qt::Tool
                   | Qt::WindowTransparentForInput);
    setAttribute(Qt::WA_TranslucentBackground);
    setAttribute(Qt::WA_ShowWithoutActivating);

    m_hideTimer = new QTimer(this);
    m_hideTimer->setSingleShot(true);
    connect(m_hideTimer, &QTimer::timeout, this, &PreviewOverlay::hidePreview);
    setupGeometry();
}

void PreviewOverlay::showPreview(int x, int y, int width, int height, const QString& text)
{
    showRegion(x, y, width, height, text);
}

void PreviewOverlay::showMousePreview(int x, int y)
{
    showClickPosition(x, y, QString());
}

void PreviewOverlay::showImagePreview(const QString& imagePath, int foundX, int foundY)
{
    m_type = Image;
    m_imagePath = imagePath;
    m_imageFoundPos = physToWidget(foundX, foundY);
    m_confidence = 0.0;
    m_lines.clear();
    startPreview();
}

void PreviewOverlay::showClickPosition(int x, int y, const QString& label)
{
    m_type = Crosshair;
    m_crosshairPos = physToWidget(x, y);
    m_text = label;
    startPreview();
}

void PreviewOverlay::showDragLine(int startX, int startY, int endX, int endY)
{
    m_type = Drag;
    m_dragStart = physToWidget(startX, startY);
    m_dragEnd = physToWidget(endX, endY);
    m_text = "Drag";
    startPreview();
}

void PreviewOverlay::showRegion(int x, int y, int width, int height, const QString& label)
{
    m_type = Region;
    const QPoint topLeft = physToWidget(x, y);
    const QPoint bottomRight = physToWidget(x + width, y + height);
    m_rect = QRect(topLeft, bottomRight).normalized();
    if (m_rect.width() < 8) {
        m_rect.setWidth(8);
    }
    if (m_rect.height() < 8) {
        m_rect.setHeight(8);
    }
    m_text = label;
    startPreview();
}

void PreviewOverlay::showScrollPosition(int x, int y, int clicks)
{
    m_type = Scroll;
    m_crosshairPos = physToWidget(x, y);
    m_scrollClicks = clicks;
    startPreview();
}

void PreviewOverlay::showImageMatch(const QString& imagePath, double confidence)
{
    m_type = Image;
    m_imagePath = imagePath;
    m_imageFoundPos = QPoint();
    m_confidence = confidence;
    m_lines.clear();
    startPreview();
}

void PreviewOverlay::showTextPreview(const QString& text, const QString& title)
{
    m_type = Text;
    m_text = title.isEmpty() ? "Text" : title;
    m_lines = QStringList{text};
    startPreview();
}

void PreviewOverlay::showHotkeyPreview(const QStringList& keys)
{
    m_type = Hotkey;
    m_text = "Hotkey";
    m_lines = QStringList{keys.join(" + ")};
    startPreview();
}

void PreviewOverlay::showActionGroupPreview(const QString& groupName, int actionCount, const QString& description)
{
    m_type = ActionGroup;
    m_text = groupName;
    m_lines = QStringList{QString("%1 actions").arg(actionCount)};
    if (!description.isEmpty()) {
        m_lines << description;
    }
    startPreview();
}

void PreviewOverlay::hidePreview()
{
    m_type = None;
    m_hideTimer->stop();
    hide();
}

void PreviewOverlay::setAutoHideDuration(int ms)
{
    m_autoHideDuration = ms;
}

void PreviewOverlay::setupGeometry()
{
    const QList<QScreen*> screens = QGuiApplication::screens();
    if (screens.isEmpty()) {
        return;
    }

    QRect virtualGeometry = screens.first()->geometry();
    for (QScreen* screen : screens) {
        virtualGeometry = virtualGeometry.united(screen->geometry());
    }
    setGeometry(virtualGeometry);
}

QPoint PreviewOverlay::physToWidget(int x, int y) const
{
    double dpr = 1.0;
    if (QScreen* screen = QGuiApplication::primaryScreen()) {
        dpr = screen->devicePixelRatio();
        if (dpr <= 0.0) {
            dpr = 1.0;
        }
    }

    const QRect geo = geometry();
    return QPoint(static_cast<int>(std::round(x / dpr)) - geo.x(),
                  static_cast<int>(std::round(y / dpr)) - geo.y());
}

void PreviewOverlay::startPreview()
{
    setupGeometry();
    show();
    raise();
    update();
    m_hideTimer->start(m_autoHideDuration);
}

void PreviewOverlay::drawLabel(QPainter& painter, const QRect& anchor, const QString& text, const QColor& color)
{
    if (text.isEmpty()) {
        return;
    }

    QFont font("Segoe UI", 10);
    painter.setFont(font);
    QFontMetrics fm(font);
    const int textWidth = std::min(fm.horizontalAdvance(text) + 16, std::max(120, width() - 24));
    const int textHeight = fm.height() + 8;
    QRect labelRect(anchor.x(), anchor.y() - textHeight - 6, textWidth, textHeight);
    if (labelRect.top() < 8) {
        labelRect.moveTop(anchor.bottom() + 6);
    }
    if (labelRect.right() > width() - 8) {
        labelRect.moveRight(width() - 8);
    }
    if (labelRect.left() < 8) {
        labelRect.moveLeft(8);
    }

    painter.setPen(Qt::NoPen);
    painter.setBrush(color);
    painter.drawRoundedRect(labelRect, 4, 4);
    painter.setPen(Qt::white);
    painter.drawText(labelRect.adjusted(8, 0, -8, 0), Qt::AlignVCenter | Qt::TextSingleLine, text);
}

void PreviewOverlay::drawInfoCard(QPainter& painter, const QString& title, const QStringList& lines, const QColor& color)
{
    QFont titleFont("Segoe UI", 12, QFont::DemiBold);
    QFont bodyFont("Segoe UI", 10);
    QFontMetrics titleFm(titleFont);
    QFontMetrics bodyFm(bodyFont);

    int contentWidth = titleFm.horizontalAdvance(title);
    for (const QString& line : lines) {
        contentWidth = std::max(contentWidth, bodyFm.horizontalAdvance(line));
    }
    const int cardWidth = std::clamp(contentWidth + 44, 260, std::max(260, width() - 48));
    const int lineCount = std::max(1, static_cast<int>(lines.size()));
    const int cardHeight = 48 + titleFm.height() + lineCount * (bodyFm.height() + 4);
    QRect card((width() - cardWidth) / 2, (height() - cardHeight) / 2, cardWidth, cardHeight);

    painter.setPen(Qt::NoPen);
    painter.setBrush(QColor(0, 0, 0, 90));
    painter.drawRect(rect());

    painter.setBrush(QColor(255, 255, 255, 242));
    painter.drawRoundedRect(card, 8, 8);
    painter.setBrush(color);
    painter.drawRoundedRect(QRect(card.left(), card.top(), 6, card.height()), 3, 3);

    QRect titleRect = card.adjusted(22, 14, -18, -card.height() + 38);
    painter.setFont(titleFont);
    painter.setPen(QColor(24, 24, 24));
    painter.drawText(titleRect, Qt::AlignVCenter | Qt::TextSingleLine, title);

    painter.setFont(bodyFont);
    painter.setPen(QColor(78, 78, 78));
    int y = titleRect.bottom() + 10;
    for (const QString& line : lines) {
        QRect lineRect(card.left() + 22, y, card.width() - 40, bodyFm.height() + 2);
        painter.drawText(lineRect, Qt::AlignVCenter | Qt::TextSingleLine, line);
        y += bodyFm.height() + 4;
    }
}

void PreviewOverlay::paintEvent(QPaintEvent*)
{
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);

    switch (m_type) {
    case Region: {
        painter.setPen(Qt::NoPen);
        painter.setBrush(QColor(0, 0, 0, 80));
        painter.drawRect(rect());

        painter.setPen(QPen(QColor(0, 120, 212), 2));
        painter.setBrush(QColor(0, 120, 212, 35));
        painter.drawRoundedRect(m_rect, 4, 4);
        drawLabel(painter, m_rect, m_text, QColor(0, 120, 212));
        break;
    }
    case Crosshair: {
        int armLen = 16;
        painter.setPen(QPen(QColor(255, 50, 50, 200), 2));
        painter.drawLine(m_crosshairPos.x() - armLen, m_crosshairPos.y(),
                         m_crosshairPos.x() + armLen, m_crosshairPos.y());
        painter.drawLine(m_crosshairPos.x(), m_crosshairPos.y() - armLen,
                         m_crosshairPos.x(), m_crosshairPos.y() + armLen);

        painter.setPen(Qt::NoPen);
        painter.setBrush(QColor(255, 50, 50, 60));
        painter.drawEllipse(m_crosshairPos, 6, 6);

        QFont font("Segoe UI", 10);
        painter.setFont(font);
        QString coordText = m_text.isEmpty()
            ? QString("(%1, %2)").arg(m_crosshairPos.x()).arg(m_crosshairPos.y())
            : m_text;
        QFontMetrics fm(font);
        int tw = fm.horizontalAdvance(coordText) + 12;
        int th = fm.height() + 6;
        QRect coordRect(m_crosshairPos.x() + 12, m_crosshairPos.y() + 12, tw, th);
        painter.setPen(Qt::NoPen);
        painter.setBrush(QColor(30, 30, 30, 200));
        painter.drawRoundedRect(coordRect, 3, 3);
        painter.setPen(Qt::white);
        painter.drawText(coordRect, Qt::AlignCenter, coordText);
        break;
    }
    case Image: {
        if (!m_imageFoundPos.isNull()) {
            int boxSize = 40;
            QRect boxRect(m_imageFoundPos.x() - boxSize / 2, m_imageFoundPos.y() - boxSize / 2,
                          boxSize, boxSize);
            painter.setPen(QPen(QColor(0, 160, 80), 3));
            painter.setBrush(QColor(0, 160, 80, 30));
            painter.drawRect(boxRect);
            drawLabel(painter, boxRect,
                      QString("(%1, %2)").arg(m_imageFoundPos.x()).arg(m_imageFoundPos.y()),
                      QColor(0, 160, 80));
        } else {
            QStringList lines;
            lines << QFileInfo(m_imagePath).fileName();
            if (m_confidence > 0.0) {
                lines << QString("confidence %1%").arg(static_cast<int>(m_confidence * 100));
            }
            drawInfoCard(painter, "Image match", lines, QColor(0, 160, 80));
        }
        break;
    }
    case Drag: {
        painter.setPen(Qt::NoPen);
        painter.setBrush(QColor(0, 0, 0, 70));
        painter.drawRect(rect());
        painter.setPen(QPen(QColor(0, 120, 212), 3, Qt::SolidLine, Qt::RoundCap));
        painter.drawLine(m_dragStart, m_dragEnd);
        painter.setBrush(QColor(0, 120, 212, 50));
        painter.drawEllipse(m_dragStart, 7, 7);
        painter.drawEllipse(m_dragEnd, 7, 7);
        QRect anchor(m_dragStart, m_dragEnd);
        drawLabel(painter, anchor.normalized(), "Drag", QColor(0, 120, 212));
        break;
    }
    case Scroll: {
        const int direction = m_scrollClicks >= 0 ? -1 : 1;
        const int length = 42;
        QPoint arrowEnd(m_crosshairPos.x(), m_crosshairPos.y() + direction * length);
        painter.setPen(QPen(QColor(196, 90, 17), 3, Qt::SolidLine, Qt::RoundCap));
        painter.drawLine(m_crosshairPos, arrowEnd);
        QPolygon arrow;
        arrow << arrowEnd
              << QPoint(arrowEnd.x() - 7, arrowEnd.y() - direction * 10)
              << QPoint(arrowEnd.x() + 7, arrowEnd.y() - direction * 10);
        painter.setBrush(QColor(196, 90, 17));
        painter.drawPolygon(arrow);
        drawLabel(painter, QRect(m_crosshairPos.x() - 20, m_crosshairPos.y() - 20, 40, 40),
                  QString("Scroll %1").arg(m_scrollClicks), QColor(196, 90, 17));
        break;
    }
    case Text:
        drawInfoCard(painter, m_text, m_lines, QColor(0, 120, 212));
        break;
    case Hotkey:
        drawInfoCard(painter, m_text, m_lines, QColor(120, 70, 180));
        break;
    case ActionGroup:
        drawInfoCard(painter, QString("Action group: %1").arg(m_text), m_lines, QColor(0, 153, 188));
        break;
    case None:
        break;
    }
}
